src/common/survival_util.cc

```diff
@@ -0,0 +1,282 @@
+  double pdf;
+  double hess;
+  double w;
+  pdf = this->PDF(z);
+  w = std::exp(z);
+  hess = pdf * (w * w - 4 * w + 1) / ((1 + w) * (1 + w));
+  return hess;
+}
+
+double AFTExtreme::PDF(double z) {
+  double pdf;
+  double w;
+  w = std::exp(z);
+  pdf = w * std::exp(-w);
+  return pdf;
+}
+
+double AFTExtreme::CDF(double z) {
+  double cdf;
+  double w;
+  w = std::exp(z);
+  cdf = 1 - std::exp(-w);
+  return cdf;
+}
+
+double AFTExtreme::GradPDF(double z) {
+  double pdf;
+  double w;
+  double grad;
+
+  pdf = this->PDF(z);
+  w = std::exp(z);
+  grad = (1 - w) * pdf;
+  return grad;
+}
+
+double AFTExtreme::HessPDF(double z) {
+  double pdf;
+  double w;
+  double hess;
+  pdf = this->PDF(z);
+  w = std::exp(z);
+  hess = (w * w - 3 * w + 1) * pdf;
+  return hess;
+}
+
+
+double AFTLoss::Loss(double y_lower, double y_upper, double y_pred, double sigma) {
+  double pdf;
+  double cdf_u, cdf_l, z_u, z_l;
+  double cost;
+
+  const double log_y_lower = std::log(y_lower);
+  const double log_y_upper = std::log(y_upper);
+  if (y_lower == y_upper) {  // uncensored
+    z_l = (log_y_lower - y_pred) / sigma;
+    pdf = dist_->PDF(z_l);
+    cost = -std::log(pdf / (sigma * y_lower));
+  } else {  // censored; now check what type of censorship we have
+    if (std::isinf(y_upper)) {  // right-censored
+      cdf_u = 1;
+    } else {  // left-censored or interval-censored
+      z_u = (log_y_upper - y_pred) / sigma;
+      cdf_u = dist_->CDF(z_u);
+    }
+    if (std::isinf(y_lower)) {  // left-censored
+      cdf_l = 0;
+    } else {  // right-censored or interval-censored
+      z_l = (log_y_lower - y_pred) / sigma;
+      cdf_l = dist_->CDF(z_l);
+    }
+    cost = -std::log(cdf_u - cdf_l);
+  }
+  return cost;
+}
+
+double AFTLoss::Gradient(double y_lower, double y_upper, double y_pred, double sigma) {
+  double pdf_l;
+  double pdf_u;
+  double pdf;
+  double grad;
+  double z;
+  double z_u;
+  double z_l;
+  double cdf_u;
+  double cdf_l;
+  double gradient;
+  const double eps = 1e-12f;
+
+  const double log_y_lower = std::log(y_lower);
+  const double log_y_upper = std::log(y_upper);
+  if (y_lower == y_upper) {  // uncensored
+    z = (log_y_lower - y_pred) / sigma;
+    pdf = dist_->PDF(z);
+    grad = dist_->GradPDF(z);
+    gradient = grad / (sigma * pdf);
+  } else {  // censored; now check what type of censorship we have
+    if (std::isinf(y_upper)) {  // right-censored
+      pdf_u = 0;
+      cdf_u = 1;
+    } else {  // interval-censored or left-censored
+      z_u = (log_y_upper - y_pred) / sigma;
+      pdf_u = dist_->PDF(z_u);
+      cdf_u = dist_->CDF(z_u);
+    }
+    if (std::isinf(y_lower)) {  // left-censored
+      pdf_l = 0;
+      cdf_l = 0;
+    } else {  // interval-censored or right-censored
+      z_l = (log_y_lower - y_pred) / sigma;
+      pdf_l = dist_->PDF(z_l);
+      cdf_l = dist_->CDF(z_l);
+    }
+    gradient = (pdf_u - pdf_l) / (sigma * std::max(cdf_u - cdf_l, eps));
```

Comment:
@avinashbarnwal I'm not sure if we should use `eps` here. It makes our C++ code disagree with the reference plot at https://github.com/avinashbarnwal/GSOC-2019/blob/master/AFT/R/combined_assignment.R.

true label = [60, +inf), predicted label = log(2), sigma = 1, dist = extreme. R code gives us gradient = -29.9950, where our C++ code gives -2.80728

Reply:
Also, this is not strictly correct, since `cdf_u - cdf_l` could be a small negative number. For negative number, we need to use `-eps` instead. That is,
`gradient =  (pdf_u - pdf_l) / (sigma * apply_eps(cdf_u - cdf_l, eps));`
where `apply_eps` is equivalent to `ifelse(x >= 0, pmax(x, eps), pmin(x, -eps)`

Reply:
@hcho3 Generally eps is used to stabilize the denominator. It is very common in regularization. Please let me know if you want any references for that.

Reply:
Yes, a reference would be nice.

Reply:
I can remember this paper off the top of the head -  https://arxiv.org/pdf/1603.06743.pdf. See page number 4 in the footnote. It is also commonly mentioned in Elements of Statistical Learning Book.

Reply:
Thanks. Let me go ahead and add a comment explaining the effect of `eps`.